### DISCUSSION TITLE
Regexp filter for message handler only match if message is 2 words

### DISCUSSION THREAD

**Discussion by 100nandoo** [aiogram/aiogram#581] [Q&A] [answered]
I want to create regexp filter for message handler but only if message is exactly 2 words?

Example message:
`BTC USDT`

**Comment by evgfilim1** [comment, marked as answer]
Hello, @100nandoo. You can use this regexp for your handler: `^(\w+) (\w+)$`

## Explanation
- `^` asserts the start of the line
- `(...)` starts capturing group, so you can use the result of this capture later
- `\w` is the same as `[a-zA-Z0-9_]`
  - `[...]` is a choice list, so it matches exactly one character from this list
  - `[a-z]` matches everything from `a` (ASCII codepoint 97) to `z` (ASCII codepoint 122)
  - `[a-z0-9]` helps you to match either `[a-z]` or `[0-9]`
  - so `[a-zA-Z0-9_]` is one of `[a-z]`, `[A-Z]`, `[0-9]` or `_`.
- `+` is the same as `{1,}`
  - In common, `{a,b}` means "match previous expression (`\w` in this case) from `a` to `b` times"
  - If `a` is missing, it is considered equal 0
  - If `b` is missing, it is considered maximum
- ` ` matches Space literally
- `$` asserts the end of the line

## Example
```python
@dp.message_handler(regexp=r"^(\w+) (\w+)$")
async def words(message: types.Message, match: typing.Match[str]):
    await message.reply(f"Convert {match[1]} to {match[2]}")  # Note that capturing groups are counted from 1 as 0 is reserved for full match
```
In your case it would reply "Convert BTC to USDT"

## See also
- https://regex101.com/ — very useful webapp to help learn regexps, debug them and find new ones
- https://docs.python.org/3/library/re.html — Python documentation for `re` module, which is used for regexps
- https://github.com/aiogram/aiogram/blob/d5a4c0c4afc6f03b9de19f7f68f852c9188cf072/aiogram/dispatcher/filters/builtin.py#L403

**Reply by 100nandoo** [reply]
Thanks for the thorough explanation. I found another issue, while I was using this regex. My other message handler that accept more than 2 words is never trigger anymore.

Is there a way to trigger it if only the word is exactly 2?

**Reply by Olegt0rr** [reply]
It's not a school of coding.
Please use google first.

Just check `len(message.text.split()) == 2` in the filter.

**Reply by 100nandoo** [reply]
Understood. I didn't know there is `filter` method inside Regexp. Thanks
